File: MATTHEW/leetcode/AIcoding/2.Gemini-2.0/cpp_solutions/LCR 113-QA2IGt.cpp
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> inDegree(numCourses, 0);
        vector<vector<int>> adj(numCourses);
        vector<int> result;

        for (auto& p : prerequisites) {
            int course = p[0];
            int pre = p[1];
            adj[pre].push_back(course);
            inDegree[course]++;
        }

        queue<int> q;
        for (int i = 0; i < numCourses; ++i) {
            if (inDegree[i] == 0) {
                q.push(i);
            }
        }

        int count = 0;
        while (!q.empty()) {
            int u = q.front();
            q.pop();
            result.push_back(u);
            count++;

            for (int v : adj[u]) {
                inDegree[v]--;
                if (inDegree[v] == 0) {
                    q.push(v);
                }
            }
        }

        if (count == numCourses) {
            return result;
        } else {
            return {};
        }
    }
};

File: leetcode/AIcoding/2.Gemini-2.0/go_solutions/LCR 113-QA2IGt.go
func findOrder(numCourses int, prerequisites [][]int) []int {
    inDegree := make([]int, numCourses)
    adj := make([][]int, numCourses)

    for _, p := range prerequisites {
        inDegree[p[0]]++
        adj[p[1]] = append(adj[p[1]], p[0])
    }

    queue := []int{}
    for i := 0; i < numCourses; i++ {
        if inDegree[i] == 0 {
            queue = append(queue, i)
        }
    }

    count := 0
    result := []int{}

    for len(queue) > 0 {
        course := queue[0]
        queue = queue[1:]
        result = append(result, course)
        count++

        for _, nextCourse := range adj[course] {
            inDegree[nextCourse]--
            if inDegree[nextCourse] == 0 {
                queue = append(queue, nextCourse)
            }
        }
    }

    if count == numCourses {
        return result
    }

    return []int{}
}